protobuf-format.md

```diff
@@ -0,0 +1,181 @@
+construct similar to that prescriped in the [JSON format][json-format].
+
+The [Attributes](#2-attributes) section describes the naming conventions and
+data type mappings for CloudEvents attributes for use as protobuf message
+properties.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Content-Type
+
+There is no official IANA *media-type* designation for protobuf, as such this
+specification uses 'application/x-protobuf' to identify such content.
+
+## 2. Type System
+
+The CloudEvents type system is mapped to protobuf as follows :
+
+| CloudEvents   | protobuf |
+| ------------- | ---------------------------------------------------------------------- |
+| Boolean       | [boolean][proto-scalars] |
+| Integer       | [sint32][proto-scalars] |
+| String        | [string][proto-scalars] |
+| Binary        | [bytes][proto-scalars] |
+| URI           | [string][proto-scalars] following [RFC 3986 §4.3][rfc3986-section43]|
+| URI-reference | [string][proto-scalars] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [Timestamp][proto-timestamp]  |
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are represented in the protobuf
+[schema][proto-schema].
+
+Required and Optional attributes are represented explicitly as protobuf fields.
+
+CloudEvent extension attributes are represented using a map construct enabling
+direct support of the CloudEvent [type system][ce-types].
+
+```proto
+map<string, CloudEventAttribute> extensions = 1;
+
+message CloudEventAttribute {
+
+    oneof attr_oneof {
+      bool ce_boolean = 1;
+      sfixed32 ce_integer = 2;
+      string ce_string = 3;
+      bytes ce_binary = 4;
+      string ce_uri = 5;
+      string ce_uri_reference = 6;
+      google.protobuf.Timestamp ce_timestamp = 7;
+    }
+}
+```
+
+NOTE: Field assignment numbers are non-normative.
+
+In this model an extension attribute's name is used as the map *key* and is
+associated with its *value* stored in the appropriately typed property.
+
+This approach allows for extensions to be represented and transported
+with no loss of *type* information.
+
+## 3. Data
+
+The specification allows for data payloads of the following types to be explicitly represented:
+
+* string
+* bytes
+* protobuf object/message
+
+```proto
+oneof data_oneof {
+
+    // Binary data
+    bytes raw_data = 2;
+
+    // String data
+    string text_data = 3;
+
+    // Protobuf Message data
+    google.protobuf.Any proto_data = 4;
+}
+```
+
+Before encoding, a protobuf serializer MUST first determine the runtime data type
+of the content. This can be determined by examining the data for invalid UTF-8
+sequences or by consulting the `datacontenttype` attribute.
+
+* Where the data is a protobuf message it MUST be stored in the `proto_data`
```

Comment:
Why complicate this vs. just using `bytes` (and possibly needing to decode the payload a second time), vs needing to scan the payload to determine which field to put the content into?

Also, is it an error to receive a utf-8 string or a proto message as `raw_data`? If not, this seems like a case where the recipient will know what to do best.

Reply:
I get your point but I'm not sure it complicates things. My thought process was that it was a bit odd to put textual content into a byte array when Proto allow us to be explicit.

Reply:
FWIW, the binary encoding of protobuf lets you "Pun" two different protos between different length-delimited field types, like so:

```proto
// Use this type if you are an intermediate router and don't need to care about interpreting or validating payload.
message Envelope {
  string name =1;
  bytes payload = 2;
}

message Data {
  map<string, int32> values = 1;
  double magnitude = 2;
}

// If you are an endpoint node, decode Envelope as Decoded and you can interpret payload as the Data message.
message Decoded {
  string name = 1;
  Data data = 2;
}
```

The main benefit of using `string` vs `bytes` is that Proto will generate the correct types from the auto-generated code, possibly verifying that the string is valid UTF-8 along the way (I don't recall if all implementations verify the UTF-8-ness of the bytestream). The cost is that any proto interpretation which understands the message will unpack and verify all the inner payload fields, whether or not they are used by the application.